File: caro_client_export/full_clean_export/backend/engine/validation/validator.py
"""
Validation against YOTEL D01-C08 brand standards and Barbados Class VI planning rules.
Returns (is_valid, violations, warnings).
"""

import math
from typing import Any, Dict, List, Tuple

from .rules import get_rules
from config.site import OFFSET_BOUNDARY, VIEWER_BUILDING_PLACEMENT


def _point_in_poly(x: float, y: float, poly: List[Dict[str, float]]) -> bool:
    """
    Ray casting point-in-polygon test.

    Args:
        x, y: point in local metres
        poly: polygon points as [{"x":..., "y":...}, ...]
    """
    def _point_on_segment(px: float, py: float, ax: float, ay: float, bx: float, by: float, eps: float = 1e-9) -> bool:
        # Check colinearity via cross-product + bounding-box containment via dot product.
        cross = (px - ax) * (by - ay) - (py - ay) * (bx - ax)
        if abs(cross) > eps:
            return False
        dot = (px - ax) * (px - bx) + (py - ay) * (py - by)
        return dot <= eps

    # Treat points exactly on the boundary as "inside".
    n = len(poly)
    for i in range(n):
        a = poly[i]
        b = poly[(i + 1) % n]
        if _point_on_segment(x, y, a["x"], a["y"], b["x"], b["y"]):
            return True

    inside = False
    for i in range(n):
        x1, y1 = poly[i]["x"], poly[i]["y"]
        x2, y2 = poly[(i - 1) % n]["x"], poly[(i - 1) % n]["y"]
        if ((y1 > y) != (y2 > y)) and (x < (x2 - x1) * (y - y1) / ((y2 - y1) or 1e-12) + x1):
            inside = not inside
    return inside


def _rotate_point(x: float, y: float, ang_rad: float) -> Tuple[float, float]:
    """2D rotation around origin by angle in radians."""
    c = math.cos(ang_rad)
    s = math.sin(ang_rad)
    return x * c - y * s, x * s + y * c


def validate(opt: Dict[str, Any]) -> Tuple[bool, List[str], List[str]]:
    """
    Validate an option against all rules.
    
    Args:
        opt: dict with "metrics" and "params" keys
    
    Returns:
        (is_valid: bool, violations: list[str], warnings: list[str])
        Violations = hard fail. Warnings = soft issues.
    """
    violations = []
    warnings = []
    m = opt["metrics"]
    p = opt.get("params", {})

    rules = get_rules()
    PLANNING = rules.get("planning", {})
    BRAND = rules.get("brand", {})

    # ── Planning rules ──────────────────────────────────
    if m["coverage_pct"] > PLANNING["max_coverage_pct"]:
        violations.append(f"Coverage {m['coverage_pct']:.0f}% > {PLANNING['max_coverage_pct']}% max")

    if m["building_height_m"] > PLANNING["max_height_m"]:
        violations.append(f"Height {m['building_height_m']:.1f}m > {PLANNING['max_height_m']}m max")

    # ── Offset boundary containment (buildable zone) ─────
    # Enforce that the generated footprint does not extend outside the offset boundary polygon.
    # Uses the same placement/rotation as the canonical viewer.
    try:
        ang = VIEWER_BUILDING_PLACEMENT.get("rot_deg", 0) * math.pi / 180
        tx = VIEWER_BUILDING_PLACEMENT.get("x", 0)
        ty = VIEWER_BUILDING_PLACEMENT.get("y", 0)
        outside = 0
        for wing in m.get("wings", []):
            x0 = wing.get("x", 0)
            y0 = wing.get("y", 0)
            # Match Viewer3D geometry rules:
            # - viewer draws EW wings as (l along X) × (w along Z)
            # - viewer draws NS wings as (w along X) × (l along Z)
            L = wing.get("l", 0)
            W = wing.get("w", 0)
            if wing.get("dir") == "NS":
                rect_Lx, rect_Wy = W, L
            else:
                rect_Lx, rect_Wy = L, W
            corners = [(x0, y0), (x0 + rect_Lx, y0), (x0 + rect_Lx, y0 + rect_Wy), (x0, y0 + rect_Wy)]
            for cx, cy in corners:
                rx, ry = _rotate_point(cx, cy, ang)
                wx, wy = rx + tx, ry + ty
                if not _point_in_poly(wx, wy, OFFSET_BOUNDARY):
                    outside += 1
        if outside > 0:
            violations.append("Building footprint exceeds the offset boundary (buildable zone)")
    except Exception:
        # keep validator resilient
        pass

    bL = m.get("bounding_L", 0)
    bW = m.get("bounding_W", 0)
    if bL + 2 * PLANNING["min_side_setback_m"] > PLANNING["site_length_m"]:
        violations.append(f"Building length {bL}m + setbacks > site length {PLANNING['site_length_m']}m")
    if bW + 2 * PLANNING["min_side_setback_m"] > PLANNING["site_width_m"]:
        violations.append(f"Building width {bW}m + setbacks > site width {PLANNING['site_width_m']}m")

    # ── Brand width rules ───────────────────────────────
    width_mm = m.get("wing_width_m", 14) * 1000
    ct = m.get("corridor_type", "double")
    if ct == "double":
        if width_mm < BRAND["dual_min_width_mm"]:
            violations.append(f"Dual-aspect needs ≥{BRAND['dual_min_width_mm']}mm; got {width_mm:.0f}mm")
        if m.get("pad_units", 0) > 0 and width_mm < BRAND["pad_dual_min_width_mm"]:
            warnings.append(f"PAD dual-aspect needs {BRAND['pad_dual_min_width_mm']}mm; building is {width_mm:.0f}mm")
    else:
        if width_mm < BRAND["single_min_width_mm"]:
            violations.append(f"Single-aspect needs ≥{BRAND['single_min_width_mm']}mm; got {width_mm:.0f}mm")

    # ── GIA/key sanity ──────────────────────────────────
    gk = m.get("gia_per_key", 0)
    if gk < 25:
        violations.append(f"GIA/key {gk}m² impossibly tight")
    elif gk < 29:
        warnings.append(f"GIA/key {gk}m² below YOTEL benchmark (29m²)")
    elif gk > 48:
        warnings.append(f"GIA/key {gk}m² — inefficient")

    # ── Accessibility ───────────────────────────────────
    total_keys = m.get("total_keys", 0)
    needed = max(1, math.ceil(total_keys * BRAND["min_accessible_pct"]))
    actual = m.get("accessible_count", 0)
    if actual < needed:
        warnings.append(f"Need {needed} accessible keys (5%); confirm during detail design")

    # ── FOH space adequacy ──────────────────────────────
    fp = m.get("footprint_m2", 0)
    foh_area = fp * p.get("ground_foh_pct", 0.60)
    min_foh = BRAND["min_komyuniti_m2"] + BRAND["min_mission_control_m2"] + BRAND["min_gym_m2"]
    if foh_area < min_foh:
        warnings.append(f"FOH area {foh_area:.0f}m² tight for Komyuniti+MC+Gym")

    # ── Lifts ───────────────────────────────────────────
    m["required_foh_lifts"] = 1 + math.ceil(total_keys / 100)
    m["required_boh_lifts"] = max(1, math.ceil(total_keys / 300))

    return len(violations) == 0, violations, warnings

File: studio/src/engine/validator.ts
import type { OptionMetrics, Wing, Point2D, ValidationResult, Violation, ProjectId } from './types'
import { RULES } from '@/config/rules'
import { SITE as CARLISLE_BAY_SITE } from '@/config/site'
import { SITE as ABBEVILLE_SITE } from '@/config/abbeville/site'
import { getJurisdiction, type Jurisdiction } from '@/config/jurisdictions'

function getProjectSite(projectId?: ProjectId) {
  if (projectId === 'abbeville') return ABBEVILLE_SITE
  return CARLISLE_BAY_SITE
}

/** Ray-casting point-in-polygon test. */
export function pointInPolygon(x: number, y: number, poly: Point2D[]): boolean {
  let inside = false
  for (let i = 0, j = poly.length - 1; i < poly.length; j = i++) {
    const xi = poly[i].x, yi = poly[i].y
    const xj = poly[j].x, yj = poly[j].y
    if ((yi > y) !== (yj > y) && x < ((xj - xi) * (y - yi)) / (yj - yi) + xi) {
      inside = !inside
    }
  }
  return inside
}

/** Rotate a 2D point around origin by angle in radians (per validator.py). */
function rotatePoint(x: number, y: number, angRad: number): [number, number] {
  const c = Math.cos(angRad), s = Math.sin(angRad)
  return [x * c - y * s, x * s + y * c]
}

/**
 * Validate a design option against planning rules.
 *
 * @param metrics  - the computed option metrics
 * @param wings    - array of building wings
 * @param jurisdictionId - optional 2-letter jurisdiction code (e.g. 'bb', 'jm').
 *                         Falls back to Barbados defaults if omitted or not found.
 * @param projectId - optional project identifier. When provided, uses that project's
 *                    site config (maxFootprint, buildableEW/NS). Defaults to carlisle-bay.
 */
export function validate(
  metrics: OptionMetrics,
  wings: Wing[],
  jurisdictionId?: string,
  projectId?: ProjectId,
): ValidationResult {
  const violations: Violation[] = []
  const warnings: string[] = []

  // Resolve site config based on projectId (defaults to carlisle-bay)
  const SITE = getProjectSite(projectId)

  // Resolve jurisdiction — fall back to Barbados hardcoded RULES when absent
  const jurisdiction: Jurisdiction | undefined = jurisdictionId
    ? getJurisdiction(jurisdictionId)
    : undefined

  // Effective planning limits: jurisdiction overrides if present, else Barbados defaults
  const maxCoverage = jurisdiction?.maxCoverage ?? RULES.planning.maxCoverage
  const maxHeight = jurisdiction?.maxHeight ?? RULES.planning.maxHeight
  const boundarySetback = jurisdiction?.minSetback ?? RULES.planning.boundarySetback
  const buildingSeparation = RULES.planning.buildingSeparation // site-specific, keep Barbados

  // Coverage
  if (metrics.coverage > maxCoverage) {
    violations.push({
      rule: 'Max site coverage',
      actual: `${(metrics.coverage * 100).toFixed(1)}%`,
      limit: `${maxCoverage * 100}%`,
      severity: 'fatal',
    })
  }

  // Height
  if (metrics.buildingHeight > maxHeight) {
    violations.push({
      rule: 'Max building height',
      actual: `${metrics.buildingHeight}m`,
      limit: `${maxHeight}m`,
      severity: 'fatal',
    })
  }

  // ── BOUNDARY SETBACK ENFORCEMENT ──────────────────────────────────────
  // Uses jurisdiction-specific setback when provided, otherwise Barbados 1.83m.
  // Wing coordinates are in building-local space (origin at BUILDING_PLACEMENT).
  // The buildable zone spans [0, buildableEW] × [0, buildableNS].
  const SETBACK = boundarySetback
  const minX = SETBACK
  const minY = SETBACK
  const maxX = SITE.buildableEW - SETBACK
  const maxY = SITE.buildableNS - SETBACK

  let outsideCorners = 0
  let tooCloseToEdge = 0
  for (const wing of wings) {
    const rectLx = wing.direction === 'NS' ? wing.width : wing.length
    const rectWy = wing.direction === 'NS' ? wing.length : wing.width
    const corners: [number, number][] = [
      [wing.x, wing.y],
      [wing.x + rectLx, wing.y],
      [wing.x + rectLx, wing.y + rectWy],
      [wing.x, wing.y + rectWy],
    ]
    for (const [cx, cy] of corners) {
      // Hard fail: completely outside buildable zone
      if (cx < 0 || cy < 0 || cx > SITE.buildableEW || cy > SITE.buildableNS) {
        outsideCorners++
      }
      // Setback violation
      else if (cx < minX || cy < minY || cx > maxX || cy > maxY) {
        tooCloseToEdge++
      }
    }
  }

  if (outsideCorners > 0) {
    violations.push({
      rule: 'Building outside buildable zone',
      actual: `${outsideCorners} corners outside boundary`,
      limit: '0 — must be fully within offset boundary',
      severity: 'fatal',
    })
  }

  if (tooCloseToEdge > 0) {
    violations.push({
      rule: `Boundary setback violation (${SETBACK}m minimum)`,
      actual: `${tooCloseToEdge} corners within ${SETBACK}m of boundary`,
      limit: `${SETBACK}m from all property boundaries`,
      severity: 'fatal',
    })
  }

  // ── BUILDING SEPARATION ──────────────────────────────────────────────
  // Minimum 12 ft (3.66m) between buildings on the same land.
  // Check gap between wings for multi-wing forms (L, U, C).
  if (wings.length >= 2) {
    const SEPARATION = buildingSeparation // 3.66m
    for (let i = 0; i < wings.length; i++) {
      for (let j = i + 1; j < wings.length; j++) {
        const a = wings[i], b = wings[j]
        // Simple gap check: if wings share an edge (e.g. L-form corner),
        // the overlap is intentional (connected). Only flag if wings are
        // detached but too close.
        const aEndX = a.x + (a.direction === 'NS' ? a.width : a.length)
        const bStartX = b.x
        const gap = bStartX - aEndX
        if (gap > 0 && gap < SEPARATION) {
          warnings.push(
            `Wings "${a.label}" and "${b.label}" are ${gap.toFixed(1)}m apart — minimum ${SEPARATION}m required between buildings`
          )
        }
      }
    }
  }

  // Wing widths
  for (const wing of wings) {
    const minWidth = metrics.corridorType === 'double_loaded'
      ? RULES.brand.dualMinWidth
      : RULES.brand.singleMinWidth ?? 8.0
    if (wing.width < minWidth) {
      violations.push({
        rule: `Min ${metrics.corridorType === 'double_loaded' ? 'dual' : 'single'}-loaded width (${wing.label})`,
        actual: `${wing.width}m`,
        limit: `${minWidth}m`,
        severity: 'fatal',
      })
    }
    if (metrics.padUnits > 0 && metrics.corridorType === 'double_loaded'
        && wing.width * 1000 < RULES.brand.padDualMinWidth * 1000) {
      warnings.push(`PAD dual-aspect needs ${RULES.brand.padDualMinWidth}m; wing is ${wing.width}m`)
    }
  }

  // GIA efficiency sanity
  if (metrics.giaPerKey < 25) {
    violations.push({
      rule: 'GIA/key impossibly tight',
      actual: `${metrics.giaPerKey.toFixed(1)}m²`,
      limit: '25m²',
      severity: 'fatal',
    })
  } else if (metrics.giaPerKey < 29 || metrics.giaPerKey > 48) {
    warnings.push(`GIA/key ${metrics.giaPerKey.toFixed(1)} m² — outside benchmark (29-48)`)
  }

  // Footprint within buildable area
  if (metrics.footprint > SITE.maxFootprint) {
    violations.push({
      rule: 'Max footprint (50% of buildable)',
      actual: `${metrics.footprint}m²`,
      limit: `${SITE.maxFootprint}m²`,
      severity: 'fatal',
    })
  }

  return {
    isValid: violations.filter(v => v.severity === 'fatal').length === 0,
    violations,
    warnings,
  }
}
